pep-0008.txt

```diff
@@ -227,8 +227,8 @@ parentheses. These should be used in preference to using a backslash
 for line continuation.
 
 Backslashes may still be appropriate at times.  For example, long,
-multiple ``with``-statements cannot use implicit continuation, so
-backslashes are acceptable::
+multiple ``with``-statements could not use implicit continuation
+before Python 3.10, so backslashes were acceptable for that case::
```

Comment:
Actually it works in CPython 3.9 too -- it was an easter egg of the PEG parser project.

Reply:
Ah, another dict-keys-are-sorted situation where it’s an implementation detail in one version and an official documented feature in the next.  Not sure if I should weigh the text down with technical correctness here!

I’ll wait to see if someone proposes to replace the whole example with something else that still requires or looks better with backslashes.

Reply:
I feel like we should give an example with parenthesized with instead. The PEP should reflect the current preferred style, not the past style.

Reply:
The context is a section about backslashes, not context managers!  This is saying that backslashes are generally discouraged, with some exceptions like multiple context managers, which now does not apply.

Reply:
From the perspective of a "regular" Python user, I suggest just picking whichever makes the more sense to tell users and go with that here.

Reply:
I don’t understand what you suggest.

Reply:
> I feel like we should give an example with parenthesized with instead. The PEP should reflect the current preferred style, not the past style.

I had the same thought initially, but I checked the context and noted what @merwok mentioned, so it makes sense to retain it unless there is a better example of backslashes (and/or until Py39 is EoL).

> I don’t understand what you suggest.

Sorry, the other comments below yours replying to Guido hadn't shown up for me when I replied; I was referring to 

> Ah, another dict-keys-are-sorted situation where it’s an implementation detail in one version and an official documented feature in the next.

From a user perspective, I was suggesting that instead of "weight[ing] down the text with technical correctness" as you mention, by making a distinction between the Python language and the CPython implementation, that instead you decide which minimum version makes most sense to tell users reading the PEP, and otherwise keep the text as it is now.

Reply:
Perhaps we could show the example with the parenthesized with first, and then note that for Python 3.9 and earlier that the backslash form is preferred, and invalid in 3.8 and earlier (although I don't think we should disadvantage PyPy or similar by documenting a CPython implementation detail in 3.9). 

However I'm not sure how great the precedent is for giving examples for different language versions in PEP 8, so feel free to ignore the suggestion.

A

Reply:
> Perhaps we could show the example with the parenthesized with first, 

Right, but as @merwok noted, the purpose of the example seems to be to demonstrate `\` usage for any use cases that require it, rather than to show how to format a `with` statement, so I'm not sure showing use of parenthesis for line continuation (which is already discussed elsewhere) is really needed here.

Reply:
Actually I like Adam’s suggestion.  The paragraph just before this one mentions parentheses as the preferred grouping tool, so we could show recommended example of context managers there then have the not-recommended backslashes shown.

Reply:
Sounds good!

Reply:
Since this section is about backslashes, what about just saying:

> Backslashes may still be appropriate for cases where implicit continuation within parentheses.

the remove the `with` statement example.  If you can't find another example that doesn't rely on `with`-statements, then enough said!

Reply:
If I understand correctly, this could potentially happen in any keyword statement with a syntax that incorporated commas to separate arguments that didn't explicitly allow commas:

* `return` and `yield` statements often have commas used within them, but that's because they are used with tuples, not have multiple distinct arguments, so they work fine
* Except statements were fixed in Python 3
* As mentioned for `with` statements were formally fixed in Python 3.10, with an undocumented implementation-dependent fix in Python 3.9
* `import` statements, unlike `from` statements, don't allow parens for continuation, but that's deliberate to not encourage users to group too many top-level imports into the same import statement, and thus would be very anti-PEP 8 anyway
* For assert statements, this is _sorta_ true, at least without some gymnastics, as described in and proposed to be fixed by [PEP 679](https://www.python.org/dev/peps/pep-0679/)

I _think_ that's everything, though I'm just a regular Pythonista and not a BDFL or FLUFL like you folks, so I could easily have missed something :)

Reply:
Yeah, I'm just gonna merge it as-is.

Reply:
I wanted to see how it looks like with Adam’s suggestion (show recommended example first, in the paragraph just before that mentions parens, then keep backslash example with amended text), but I didn’t have time to do that immediately and the discussion did not wait until I came back 🙂